File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/graph/Trigon.java
package jokatu.game.games.uzta.graph;

import ophelia.collections.set.HashSet;
import org.ejml.simple.SimpleMatrix;
import org.jetbrains.annotations.NotNull;

import java.util.Set;

public class Trigon {

	private final Node Δ1;
	private final Node Δ2;
	private final Node Δ3;
	private final Set<LineSegment> edges;
	private final Set<Node> nodes;

	public Trigon(@NotNull Node Δ1, @NotNull Node Δ2, @NotNull Node Δ3) {

		double anticlockwise = (Δ2.getX() - Δ1.getX()) * (Δ3.getY() - Δ1.getY())
		                            - (Δ2.getY() - Δ1.getY()) * (Δ3.getX() - Δ1.getX());
		this.Δ1 = Δ1;
		if (anticlockwise > 0) {
			this.Δ2 = Δ2;
			this.Δ3 = Δ3;
		} else {
			this.Δ2 = Δ3;
			this.Δ3 = Δ2;
		}

		nodes = new HashSet<Node>() {{
			add(Δ1);
			add(Δ2);
			add(Δ3);
		}};

		edges = new HashSet<LineSegment>() {{
			add(new LineSegment(Δ1, Δ2));
			add(new LineSegment(Δ2, Δ3));
			add(new LineSegment(Δ3, Δ1));
		}};
	}

	public Trigon(@NotNull LineSegment edge, @NotNull Node Δ3) {
		this(edge.getFirst(), edge.getSecond(), Δ3);
	}

	public Set<LineSegment> getEdges() {
		return edges;
	}

	public Set<Node> getNodes() {
		return nodes;
	}

	public boolean circumcircleContains(@NotNull Node node) {
		SimpleMatrix circumcircle = new SimpleMatrix(4, 4, true,
				Δ1.getX(), Δ1.getY(), Δ1.getX() * Δ1.getX() + Δ1.getY() * Δ1.getY(), 1,
				Δ2.getX(), Δ2.getY(), Δ2.getX() * Δ2.getX() + Δ2.getY() * Δ2.getY(), 1,
				Δ3.getX(), Δ3.getY(), Δ3.getX() * Δ3.getX() + Δ3.getY() * Δ3.getY(), 1,
				node.getX(), node.getY(), node.getX() * node.getX() + node.getY() * node.getY(), 1
		);
		return circumcircle.determinant() > 0;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		Trigon trigon = (Trigon) o;

		if (!Δ1.equals(trigon.Δ1)) return false;
		if (!Δ2.equals(trigon.Δ2)) return false;
		return Δ3.equals(trigon.Δ3);

	}

	@Override
	public int hashCode() {
		int result = Δ1.hashCode();
		result = 31 * result + Δ2.hashCode();
		result = 31 * result + Δ3.hashCode();
		return result;
	}
}

File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/graph/Trigon.kt
package jokatu.game.games.uzta.graph

import org.ejml.simple.SimpleMatrix

/**
 * A trigon (AKA triangle) has three points.
 */
class Trigon(
		/**
		 * The first point of the trigon.
		 */
		private val Δ1: Node,
		Δ2: Node,
		Δ3: Node
) {
	/**
	 * The second point of the trigon.  It is the next one anticlockwise (about the centre) from [Δ1].
	 */
	private val Δ2: Node

	/**
	 * The third point of the trigon.  It is the next one anticlockwise (about the centre) from [Δ2], and therefore the
	 * next one clockwise from [Δ1].
	 */
	private val Δ3: Node

	/**
	 * The set of three edges of this trigon.
	 */
	val edges: Set<LineSegment>

	/**
	 * The set of three points of this trigon.
	 */
	val nodes: Set<Node>

	constructor(edge: LineSegment, Δ3: Node) : this(edge.first, edge.second, Δ3)

	init {
		val anticlockwise = (Δ2.x - Δ1.x) * (Δ3.y - Δ1.y) - (Δ2.y - Δ1.y) * (Δ3.x - Δ1.x)
		if (anticlockwise > 0) {
			this.Δ2 = Δ2
			this.Δ3 = Δ3
		} else {
			this.Δ2 = Δ3
			this.Δ3 = Δ2
		}

		nodes = setOf(Δ1, Δ2, Δ3)

		edges = setOf(LineSegment(Δ1, Δ2), LineSegment(Δ2, Δ3), LineSegment(Δ3, Δ1))
	}

	/**
	 * Does the circumcircle of this trigon contain the given node?
	 */
	fun circumcircleContains(node: Node): Boolean {
		val circumcircle = SimpleMatrix(4, 4, true,
				Δ1.x, Δ1.y, Δ1.x * Δ1.x + Δ1.y * Δ1.y, 1.0,
				Δ2.x, Δ2.y, Δ2.x * Δ2.x + Δ2.y * Δ2.y, 1.0,
				Δ3.x, Δ3.y, Δ3.x * Δ3.x + Δ3.y * Δ3.y, 1.0,
				node.x, node.y, node.x * node.x + node.y * node.y, 1.0
		)
		return circumcircle.determinant() > 0
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (javaClass != other?.javaClass) return false

		other as Trigon

		if (Δ1 != other.Δ1) return false
		if (Δ2 != other.Δ2) return false
		if (Δ3 != other.Δ3) return false

		return true
	}

	override fun hashCode(): Int {
		var result = Δ1.hashCode()
		result = 31 * result + Δ2.hashCode()
		result = 31 * result + Δ3.hashCode()
		return result
	}
}
